File: framework/languages/java/samples/java/DeliveryDispatch/Server/DispatchCenter/src/main/java/systems/zlink/samples/deliverydispatch/server/dispatchcenter/handlers/AssignDeliveryHandler.java
package systems.zlink.samples.deliverydispatch.server.dispatchcenter.handlers;

import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.deliverydispatch.server.dispatchcenter.DispatchWorkQueue;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

@ZLinkHandlerGroup("dispatch")
public final class AssignDeliveryHandler
    implements ZLinkRequestHandler<Messages.AssignDelivery, Messages.AssignDeliveryRes> {
    private final DispatchWorkQueue queue;

    public AssignDeliveryHandler(DispatchWorkQueue queue) {
        this.queue = queue;
    }

    @Override
    public Messages.AssignDeliveryRes handle(
        Messages.AssignDelivery request,
        ZLinkRequestContext context) {
        queue.enqueue(request);
        System.err.printf(
            "deliverydispatch dispatch: queued delivery=%s customer=%s%n",
            request.deliveryId(),
            request.customerId());
        return new Messages.AssignDeliveryRes(request.deliveryId(), "pending");
    }
}

File: framework/languages/java/samples/kotlin/DeliveryDispatch/Server/DispatchCenter/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/dispatchcenter/handlers/AssignDeliveryHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.dispatchcenter.handlers

import systems.zlink.framework.channels.ZLinkRequestContext
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.deliverydispatch.server.dispatchcenter.DispatchWorkQueue
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.AssignDeliveryReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.AssignDeliveryRes

@ZLinkHandlerGroup("dispatch")
class AssignDeliveryHandler(
    private val queue: DispatchWorkQueue,
) : ZLinkSuspendingRequestHandler<AssignDeliveryReq, AssignDeliveryRes> {
    override suspend fun handle(
        request: AssignDeliveryReq,
        context: ZLinkRequestContext,
    ): AssignDeliveryRes {
        queue.enqueue(request)
        System.err.println(
            "deliverydispatch dispatch: queued delivery=${request.deliveryId} customer=${request.customerId}",
        )
        return AssignDeliveryRes(request.deliveryId, "pending")
    }
}
